lesson3.md

```diff
@@ -0,0 +1,841 @@
+<!-- .slide: id="lesson3" -->
+
+# Basic Frontend - Spring 2020
+
+Lesson 3, Tuesday, 2020-09-01
+
+
+---
+
+
+### what we learned last time
+
+* Variables are pointers to values
+
+* We can use operators with variables just like with values
+
+* Use `<script src="xxx.js"></script>` to add a JavaScript file to your HTML webpage.
+
+* Use `//` to write comments.
+
+* Use `console.log(...)` to output values to the console.
+
+
+---
+
+
+### We had a HomeWork
+
+You have an online shop, which sells hats and t-shirts. One of the customers wants to buy 2 hats and 3 t-shirts.
+
+A hat costs 3.99€ and a t-shirt costs 9.99€. What is the total cost?
+
+Create all the variables needed, and output the final result to the console.
+
+
+---
+
+
+### Homework solution
+
+```js
+let hatPrice = 3.99;
+let tShirtPrice = 9.99;
+
+let hatAmount = 2;
+let tShirtAmount = 3;
+
+let totalHatPrice = hatAmount * hatPrice;
+let totalTShirtPrice = tShirtAmount * tShirtPrice;
+
+let totalCost = totalHatPrice + totalTShirtPrice;
+
+console.log(totalCost);
+```
+
+
+---
+
+
+### Recap,  What does it print?
+
+```js
+let name = "Sevtap";
+console.log("Hello " + name);
+```
+
+Answer: It prints "Hello Sevtap" <!-- .element: class="fragment" -->
+
+
+---
+
+
+### What does it print?
+
+```js
+let greeting = "Hello";
+let name = "Sevtap";
+console.log(greeting + " " + firstName);
+```
+
+Answer: Error, "firstName" is not a variable <!-- .element: class="fragment" -->
+
+
+---
+
+
+### What does it print?
+
+```js
+console.log("It is " + temperature + " degrees");
+let temperature = 22;
+```
+
+Answer: Error, "temperature" is not available <!-- .element: class="fragment" -->
+
+
+---
+
+
+What do you think `x` and `y` are at the end of the program? What are the two possible outcomes?
+
+```js
+let x = 42;
+let y = x;
+x = 43;
+console.log(x);
+console.log(y);
+```
+
+Answer: x is 43, y is 42 <!-- .element: class="fragment" -->
+
+So, Why?  <!-- .element: class="fragment" -->
+
+
+---
+
+
+### Bass by value
+
+* This is called *pass by value*.
+* All basic data types (number, string, ...) are passed by value.
+* If you assign a variable to another variable, its value is __copied__.
```

Comment:
Is it?

earlier, you said "Variables are pointers to values". Is the value or the pointer copied? (genuinely asking)

Reply:
We say *copied* to make it easier for students to understand, but yes, we mean that the value is copied because how pointer work is beginner friendly.

in more detail:
`numbers` are always copied.
`null` and `undefined` are immutable and cannot by modified.
`strings` are immutable, and two variables can point to the same string, but any modification would create a copy.

`objects` and `arrays` are **not** copied, instead passed by reference, but we are going to mention that when we introduce them.

Reply:
I was only pointing it out because the pointer analogy was used earlier.
Isn't it easier if you say a variable is a reference to a value? Then you wouldn't need to make a distinction between the different datatypes. The examples you used before work the same way with arrays and objects.

Another thought that just came to mind: Is "pass by value" even the correct concept here. the pass-by-xy things are usually used to explain how arguments are passed to functions. Which is not happening here.